example/lib/example/shoehorn_handler.ex

```diff
@@ -0,0 +1,31 @@
+defmodule Example.RestartHandler do
+  use Shoehorn.Handler
+
+  def init(_opts) do
+    {:ok, %{restart_counts: 0}}
+  end
+
+  def application_started(app, s) do
+    IO.puts "Application started: #{inspect app}"
+    {:continue, s}
+  end
+
+  def application_exited(app, _reason, %{restart_counts: restart_counts} = state)
+      when restart_counts < 2 do
+    IO.puts("Application stopped: #{inspect(app)} #{inspect(state)}")
+    Application.ensure_all_started(app)
+    {:continue, %{state | restart_counts: restart_counts + 1}}
+  end
+
+  def application_exited(app, _reason, %{restart_counts: restart_counts} = state)
+      when restart_counts < 4 do
+    IO.puts("Application stopped: #{inspect(app)} #{inspect(state)}")
+    Application.ensure_all_started(app)
```

Comment:
How does it work if you want to delay the restart of the app after a cool down period?

Reply:
What about something like this?

```elixir
def application_exited(app, _reason, %{restart_counts: restart_counts} = state)
   :timer.apply_after(cool_down, Application, :ensure_all_started, [app])
   {:continue, %{state | restart_counts: restart_counts + 1}}
end
```

Reply:
That would work. A `spawn` with a sleep could work. There are quite a few things that could work. I think that the handler sending back a handling strategy is a good idea. It would encapsulate what we want to be done and let the shoehorn system call it.